R/global_search.R

```diff
@@ -1,30 +1,36 @@
 #' GlobalSearch.search
 #'
-#' Performs Global Search based on provided filters.
-#' @param filters (optional) Query filters.
+#' Performs Global Search based on the provided filters.
+#' Returns full API response (containing attributes: results, vaults, subjects, subjects_count, total, took and offset)
+#' @param filters (optional) Low-level filter specification.
+#' @param entities (optional) Low-level entity specification
 #' @param env (optional) Custom client environment.
+#' @param ... (optional) Additional query parameters (e.g. query, limit, offset).
 #'
 #' @examples \dontrun{
 #' # No filters are applied
 #' GlobalSearch.search()
 #'
 #' # Global Beacon search
-#' GlobalSearch.search('{"entities":[["gene","BRCA2"]]}')
+#' GlobalSearch.search(entities = '[["gene","BRCA2"]]')
 #'
 #' # Type filter (only vaults)
-#' GlobalSearch.search('{"filters":[{"and":[["type__in",["vault"]]]}]}')
+#' GlobalSearch.search(filters = '[{"and":[["type__in",["vault"]]]}]')
 #'
 #' # Advanced search
-#' GlobalSearch.search('{"filters":[{"and":[{"and":[["created_at__range",["2020-12-27","2021-12-27"]]]}]}],"query":"fuji"}')
+#' GlobalSearch.search(query = "fuji")
 #'
+#'
+#' # Multiple filters and entities
+#' GlobalSearch.search(entities = '[["gene","BRCA2"]]', filters = '[{"and":[{"and":[["created_at__range",["2021-11-28","2021-12-28"]]]},["type__in",["dataset"]]]}]')
 #' }
 #'
 #' @references
 #' \url{https://docs.solvebio.com/}
 #'
 #' @export
-GlobalSearch.search <- function(filters, env = solvebio:::.solveEnv) {
-  body = list()
+GlobalSearch.search <- function(filters, entities, env = solvebio:::.solveEnv, ...) {
```

Comment:
I would think about the function names. Which one will the user use? search() or results()?
Also, maybe make `query` the first argument here? It maps to the advanced query option on the frontend. Then you can do something like this: `GlobalSearch.search("xyz")`

Reply:
Parameter `query` is added as the first parameter, so it can be used as a positional argument as well.

Yes, I also feel like the function names can be confusing for the users and decide which ones to choose from. I've been thinking about potential renaming, here would be my suggestions:
- For the `search` function
  ```
   GlobalSearch.detailed_response
   GlobalSearch.full_response
   GlobalSearch.detailed_search
   ```
-  For the `results`/`subjects` function:
   ```
   GlobalSearch.results
   GlobalSearch.search_results
   GlobalSearch.only_results

   GlobalSearch.subjects
   GlobalSearch.search_subjects
   GlobalSearch.only_subjects
   ```

Reply:
I would focus in on 2 functions, just 1 for search and 1 for subjects. Ideally `search` would have the pagination logic in it. Maybe the current `search` function should be called `request` to signify that it handles the low level search request. The function's docs should tell the user to use `GlobalSearch.search` instead of `request`.

Reply:
I think you can simplify the Rmd examples by just combining the examples from GlobalSearch.results and GlobalSearch.search into 1 section for GlobalSearch.search. What do you think?

Reply:
Functions are refactored, so we have two main functions that users should 99% of the time:
- `GlobalSearch.search` - returns a data frame with result objects (has pagination logic in it)
- `GlobalSearch.subjects` - returns data frame with subjects

For other use cases, they may use the `GlobalSearch.request` function where they will be returned a full API response structure.

`Rmd` file has been updated with new examples and refactored a bit to accommodate new changes.

Reply:
Great thanks